network/lwip2transport/device.go

```diff
@@ -0,0 +1,191 @@
+// use the [Read] or [WriteTo] methods at a time.
+//
+// [lwIP library]: https://savannah.nongnu.org/projects/lwip/
+type LwIPTransportDevice interface {
+	network.IPDevice
+}
+
+// Compilation Guard Against Interface Implementation
+var _ network.IPDevice = (*lwIPDevice)(nil)
+var _ LwIPTransportDevice = (*lwIPDevice)(nil)
+
+type lwIPDevice struct {
+	tcp   *tcpHandler
+	udp   *udpHandler
+	stack lwipLib.LWIPStack
+
+	// whether the device has been closed
+	done chan struct{}
+
+	// async read call and its result
+	rdBuf chan []byte
+	rdN   chan int
+}
+
+// NewDevice creates a new LwIPTransportDevice. This device uses a [transport.StreamDialer] `sd` to handle TCP streams
+// and a [transport.PacketListener] `pl` to handle UDP packets.
+//
+// You can only have one active LwIPTransportDevice per process. If you try to create more than one, the behavior is
+// undefined. However, it's OK to [Close] an existing LwIPTransportDevice and then create a new one.
+func NewDevice(sd transport.StreamDialer, pl transport.PacketListener) (LwIPTransportDevice, error) {
+	if sd == nil || pl == nil {
+		return nil, errors.New("both sd and pl are required")
+	}
+
+	d := &lwIPDevice{
+		tcp:   newTCPHandler(sd),
+		udp:   newUDPHandler(pl, 30*time.Second),
+		stack: lwipLib.NewLWIPStack(),
+		done:  make(chan struct{}),
+		rdBuf: make(chan []byte),
+		rdN:   make(chan int),
+	}
+	lwipLib.RegisterTCPConnHandler(d.tcp)
+	lwipLib.RegisterUDPConnHandler(d.udp)
+	lwipLib.RegisterOutputFn(d.writeResponse)
+
+	return d, nil
+}
+
+// Close implements [io.Closer] and [network.IPDevice]. It closes the device, rendering it unusable for I/O.
+//
+// Close does not close other objects that are passed to this device, such as the [transport.StreamDialer],
+// [transport.PacketListener] or [io.Writer]. You are responsible for closing these objects yourself.
+func (d *lwIPDevice) Close() error {
+	close(d.done)
+	return d.stack.Close()
+}
+
+// MTU implements [network.IPDevice]. It returns the maximum buffer size of a single IP packet that can be processed by
+// this device.
+func (d *lwIPDevice) MTU() int {
+	return packetMTU
+}
+
+// writeResponse writes an incoming IP packet response `b` to this device. The packet can be read by calling the [Read]
+// function, or it can be redirected to an [io.Writer] if the [WriteTo] function has been called. writeResponse blocks
+// until the packet is successfully consumed by a [Read] or [WriteTo].
+//
+// writeResponse can be used as an output function for lwIP.
+//
+// writeResponse might be called by multiple goroutines (for example, when multiple UDP packets arrive at the same
+// time). We sequentialize the calls by using channels, if performance issues arise in the future, we can use other
+// more performant but more error-prone methods (e.g. the [sync] package) to resolve them.
+func (d *lwIPDevice) writeResponse(b []byte) (int, error) {
```

Comment:
Is seems like go-tun2socks completely ignores the returned values from the output function!
https://github.com/eycorsican/go-tun2socks/blob/master/core/output_export.go

I think the write behavior here is to block the write until the packet is fully read.
In that case you would loop until n == len(b) or done.

As a side note, we should probably fix `output_export.go` to properly handle errors.

Reply:
As discussed, with packets it's just a single read or write, and we replicate the UDP behavior of writing as much as we can to the buffer.